src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -68,6 +105,42 @@ int MPIDI_XPMEM_mpi_init_hook(int rank, int size, int *n_vcis_provided, int *tag
     /* Initialize other global parameters */
     MPIDI_XPMEM_global.sys_page_sz = (size_t) sysconf(_SC_PAGESIZE);
 
+    void *coop_counter_vaddr = NULL;
+    OPA_int_t *coop_counter = NULL;
+    void *rank0_coop_counter_addr = NULL;
+    /* Only rank 0 need to actually allocate an array to store counters and
+     * other processes should attach to this array through MPIDI_XPMEM_seg_regist */
+    if (!MPIR_Process.local_rank) {
+        coop_counter =
+            (OPA_int_t *) MPL_aligned_alloc(MPIDI_XPMEM_global.sys_page_sz,
+                                            num_local * num_local * sizeof(OPA_int_t),
+                                            MPL_MEM_OTHER);
+        MPIR_ERR_CHKANDJUMP(coop_counter == NULL, mpi_errno, MPI_ERR_OTHER, "**nomem");
+
+        for (i = 0; i < num_local * num_local; ++i)
+            OPA_store_int(&coop_counter[i], 0);
+    }
```

Comment:
I discuss with Shintaro, he is not sure if OPA may not be portable after he implements OPA with `__atomic` builtins supported by compilers since interprocess atomic operation is not guaranteed.  If we hope to make it fully portable, we should emulate atomic operation using `mutex` or we can write hardware assembly to manually guarantee atomics among processes. For now, we can use OPA as a temporary solution since ch3 lock-free queue is using OPA as well. We can decide whether we need to add more support to OPA for our interprocess atomic operations after Shintaro implements new version.